FixMath2.h

```diff
@@ -0,0 +1,845 @@
+#define MAX(N1,N2) (N1 > N2 ? N1 : N2)
+
+// Experiments
+/*#define NBITSREAL(X,N) (abs(X) < (1<<N) ? N : NBITSREAL2(X,N+1))
+  #define NBITSREAL2(X,N) (abs(X) < (1<<N) ? N : NBITSREAL(X,N+1))
+  #define UFixAuto(X) (UFixMath2<NBITSREAL(X,0),0>(X))
+  #define UFixAuto(X) (UFixMath2<NBITSREAL(X,0),0>(X))*/
+
+/*
+  template<typename T>
+  constexpr byte nBitsReal(T x, byte n=0) {
+  if (abs(x) < ((T)1 << n)) {
+  return n;
+  } else {
+  return nBitsReal(x, n + 1);
+  }
+  }
+
+ #define UFixAuto(X) (UFixMath2<nBitsReal(X),0>(X))
+*/
+
+ 
+
+
+
+
+// Forward declaration
+template<byte NI, byte NF>
+class SFixMath2;
+
+
+
+/** Instanciate an unsigned fixed point math number.
+@param NI The number of bits encoding the integer part
+@param NF The number of bits encoding the fractional part
+*/
+template<byte NI, byte NF=0> // NI and NF being the number of bits for the integral and the fractionnal parts respectively.
+class UFixMath2
+{
+  static_assert(NI+NF<=64, "The total width of a UFixMath2 cannot exceed 64bits");
+  typedef typename IntegerType<((NI+NF-1)>>3)+1>::unsigned_type internal_type ; // smallest size that fits our internal integer
+  typedef typename IntegerType<((NI+NF)>>3)+1>::unsigned_type next_greater_type ; // smallest size that fits 1<<NF for sure (NI could be equal to 0). It can be the same than internal_type in some cases.
+  
+public:
+  /** Constructor
+   */
+  UFixMath2() {;}
+
+  /** Constructor from a positive floating point value.
+@param fl Floating point value
+@return An unsigned fixed point number
+  */
+  UFixMath2(float fl)  {internal_value = /*static_cast<internal_type>*/(fl * (next_greater_type(1) << NF));}
+
+  /** Constructor from a floating point value.
+@param fl Floating point value
+@return An unsigned fixed point number
+  */
+  UFixMath2(double fl)  {internal_value = static_cast<internal_type> (fl * (next_greater_type(1) << NF)); }
+  
+  /* Constructor from integer type (as_frac = false) or from fractionnal value (as_frac=true) can be used to emulate the behavior of for instance Q8n0_to_Q8n8 */
+
+    /** Constructor from an integer value which can be interpreted as both a resulting fixed point 
+math number with a fractional part equals to 0, or as a number with decimal, ie as the underlying type behind the fixed point math number.
+@param value Integer value
+@param as_raw=false with false value will be interpreted as an integer, with true it will be interpreted as a number with decimals.
+@return An unsigned fixed point number
+  */
+  template<typename T>
+  UFixMath2(T value,bool as_raw=false)
+  {
+    if (as_raw) fromRaw(value);
+    else internal_value = (internal_type(value) << NF);
+  }
+
+
+  /** Set the internal value of the fixed point math number.
+@param raw The new internal value.
+  */  
+  template<typename T>
+  void fromRaw(T raw) { internal_value = raw; }
```

Comment:
Can this be made static? That would allow easier use:

```
auto a = UFixMath<x,y>::fromRaw(z);
// rather than
UFixMath<x,y> a;
a.fromRaw(z);
```

Reply:
Oh yes, good catch. I have to say that `static` is not my most familiar C++ concept so I usually do not think of it. Will do.

Reply:
(Static is also renowned as one of the more confusing concepts of C++, esp. since the keyword has three totally different meanings, depending on where it is used.)